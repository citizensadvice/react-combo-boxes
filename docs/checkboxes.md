# `<Checkboxes>`

Generates a group of checkboxes.

Although checkboxes are a practical alternative to a using a `<Select multiple />`
and behave very like a list box.

## Usage

This is a stateless controlled component. You must respond to `onValue` to update the selected value.

```js
const [values, setValues] = useState(initialValues);

<Checkboxes
  options={options}
  values={values}
  onValues={setValues}
/>;
```

## Props

### `name: String`

Optional. `name` attribute for each checkbox.

Use the `html` option on an option if any checkbox needs a unique name.

### `options: Array<any>`

The options to display. [See options][options].

Checkboxes take an additional `description` property that allows an optional description to be added to a checkbox.

### `mapOption: (option: any) => Option | string`

Map an option into label, value, disabled and group. [See options][options]

### `values: Array<any>`

The currently selected option

### `onChange: (event: Event) => void`

Change event added to each checkbox.

### `onValues: (option: Array<any>) => void`

Called with the selected option when a user selects an option.

#### `classPrefix: string = "react-combo-boxes-checkbox'`

By default the component is built using BEM style class names. This sets the prefix for all those classes.

Set to falsey to remove the BEM classes.

#### `groupClassPrefix: string = "react-combo-boxes-checkbox-group'`

BEM class prefix for groups.

Set to falsey to remove the BEM classes.

#### render props

A number of render methods are provided to customise the appearance of the component.

Each render method has the signature

```js
(
  // The props set on the <ComboBox>
  props: object,
  // The state of the component
  state: {
    // Current option
    option: Option;
    // Is the current option selected
    checked: boolean;
    // Current group
    group: Group;
  },
  // The props passed into the component
  componentProps: object;
) => React.ReactNode
```

The render functions available are:

| Name                 | Default value                                         | Description                                                                    |
| -------------------- | ----------------------------------------------------- | ------------------------------------------------------------------------------ |
| `renderWrapper`      | `({ key, ...props }) => <div key={key} {...props} />` | Renders the wrapper for each checkbox                                          |
| `renderInput`        | `(props) => <input {...props} />`                     | Renders the checkbox input                                                     |
| `renderLabel`        | `(props) => <label {...props} />`                     | Renders label for a checkbox                                                   |
| `renderLabelWrapper` | `(props) => <Fragment {...props} />`                  | Allows an element to wraps the label and description                           |
| `renderDescription`  | `(props) => <div {...props} />`                       | Renders the optional description for checkbox                                  |
| `renderGroup`        | `({ key, ...props }) => <div key={key} {...props} />` | Wraps a group of options                                                       |
| `renderGroupLabel`   | `(props) => <div {...props} />`                       | Renders the visible label for a group. This will be ignored by a screen-reader |

[options]: options.md
