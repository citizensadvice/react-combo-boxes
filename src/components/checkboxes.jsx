import { Fragment, useCallback, useRef, memo } from 'react';
import PropTypes from 'prop-types';
import { useNormalisedOptions } from '../hooks/use_normalised_options';
import { renderGroupedOptions } from '../helpers/render_grouped_options';
import { classPrefix as defaultClassPrefix } from '../constants/class_prefix';
import { makeBEMClass } from '../helpers/make_bem_class';
import { visuallyHiddenClassName } from '../constants/visually_hidden_class_name';

export const Checkboxes = memo((rawProps) => {
  const optionisedProps = Object.freeze({
    ...rawProps,
    ...useNormalisedOptions(rawProps),
  });

  const {
    classPrefix,
    groupClassPrefix,
    name,
    onChange,
    onValues,
    options,
    renderWrapper,
    renderInput,
    renderLabel,
    renderLabelWrapper,
    renderDescription,
    renderGroup,
    renderGroupLabel,
    renderGroupAccessibleLabel,
    selectedOptions,
  } = optionisedProps;

  const checkboxRef = useRef(new Map());

  const handleRef = useCallback((identity, el, ref) => {
    if (el) {
      checkboxRef.current.set(identity, el);
    } else {
      checkboxRef.current.delete(identity);
    }

    // Allow combination with a passed ref
    if (typeof ref === 'function') {
      ref(el);
    } else if (ref) {
      ref.current = el; // eslint-disable-line no-param-reassign
    }
  }, []);

  function handleChange(e) {
    // When getting the checked values,
    // we need to get all the current checked values directly from the DOM
    // If we only use the event target and combine with this selectedOptions
    // we risk loosing values if selectedOptions is slow to update

    const identities = [];
    checkboxRef.current.forEach((value, key) => {
      if (value.checked) {
        identities.push(key);
      }
    });
    onValues?.(
      options
        .filter((o) => identities.includes(o.identity))
        .map((o) => o.value),
    );
    onChange?.(e);
  }

  return renderGroupedOptions({
    options,
    renderGroup(group) {
      const { children, key, label, html } = group;
      return renderGroup(
        {
          key,
          className: makeBEMClass(groupClassPrefix),
          children: (
            <>
              {renderGroupLabel(
                {
                  children: label,
                  className: makeBEMClass(groupClassPrefix, 'label'),
                  ...html,
                },
                { group },
                optionisedProps,
              )}
              {children}
            </>
          ),
        },
        { group },
        optionisedProps,
      );
    },
    renderOption(option) {
      const { identity, label, key, html, disabled, description, group } =
        option;
      const checked = selectedOptions.some(
        (item) => item.identity === identity,
      );

      return renderWrapper(
        {
          className: makeBEMClass(classPrefix),
          children: (
            <>
              {renderInput(
                {
                  'aria-describedby': description ? `${key}_description` : null,
                  checked,
                  disabled,
                  id: key,
                  name,
                  onChange: handleChange,
                  type: 'checkbox',
                  value: identity,
                  className: makeBEMClass(classPrefix, 'input'),
                  ...html,
                  ref: (el) => handleRef(identity, el, html?.ref),
                },
                { option, checked },
                optionisedProps,
              )}
              {renderLabelWrapper({
                children: (
                  <>
                    {renderLabel(
                      {
                        htmlFor: key,
                        children: (
                          <>
                            {group
                              ? renderGroupAccessibleLabel(
                                  {
                                    className: visuallyHiddenClassName,
                                    children: `${group.label} `,
                                  },
                                  { group },
                                  optionisedProps,
                                )
                              : null}
                            {label}
                          </>
                        ),
                        className: makeBEMClass(classPrefix, 'label'),
                      },
                      { option, checked },
                      optionisedProps,
                    )}
                    {description && ' '}
                    {!!description &&
                      renderDescription(
                        {
                          id: `${key}_description`,
                          children: description,
                          className: makeBEMClass(classPrefix, 'description'),
                        },
                        { option, group, checked },
                        optionisedProps,
                      )}
                  </>
                ),
              })}
            </>
          ),
          key,
        },
        { option, group, checked },
        optionisedProps,
      );
    },
  });
});

Checkboxes.propTypes = {
  classPrefix: PropTypes.string,
  groupClassPrefix: PropTypes.string,
  id: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onValues: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.any).isRequired,
  values: PropTypes.arrayOf(PropTypes.any),
  renderWrapper: PropTypes.func,
  renderInput: PropTypes.func,
  renderLabel: PropTypes.func,
  renderLabelWrapper: PropTypes.func,
  renderDescription: PropTypes.func,
  renderGroup: PropTypes.func,
  renderGroupLabel: PropTypes.func,
  renderGroupAccessibleLabel: PropTypes.func,
};

Checkboxes.defaultProps = {
  classPrefix: `${defaultClassPrefix}checkbox`,
  groupClassPrefix: `${defaultClassPrefix}checkbox-group`,
  onValues: undefined,
  onChange: undefined,
  values: null,
  renderWrapper: (props) => <div {...props} />,
  renderInput: (props) => <input {...props} />,
  renderLabel: (props) => <label {...props} />,
  // eslint-disable-next-line react/jsx-no-useless-fragment
  renderLabelWrapper: (props) => <Fragment {...props} />,
  renderDescription: (props) => <div {...props} />,
  renderGroup: (props) => <div {...props} />,
  renderGroupLabel: (props) => <div {...props} />,
  renderGroupAccessibleLabel: (props) => <span {...props} />,
};

Checkboxes.displayName = 'Checkboxes';
