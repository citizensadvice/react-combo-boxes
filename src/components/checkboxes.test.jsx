import { useState, createRef } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Checkboxes } from './checkboxes';

function Test({ values: _values, ...props }) {
  const [values, setValues] = useState(_values);

  return (
    <Checkboxes
      id="id"
      values={values}
      onValues={setValues}
      {...props}
    />
  );
}

describe('options', () => {
  describe('as array of strings', () => {
    const options = ['Apple', 'Banana', 'Orange'];

    it('renders checkbox group', () => {
      const { container } = render(<Test options={options} />);
      expect(container).toMatchSnapshot();
    });

    it('renders checkboxes with selected values', () => {
      render(
        <Test
          options={options}
          values={['Apple', 'Orange']}
        />,
      );
      expect(screen.getAllByRole('checkbox', { checked: true })).toEqual([
        screen.getByRole('checkbox', { name: 'Apple' }),
        screen.getByRole('checkbox', { name: 'Orange' }),
      ]);
    });

    it('triggers the onValues callback', async () => {
      const spy = jest.fn();
      render(
        <Test
          options={options}
          onValues={spy}
          values={['Apple']}
        />,
      );
      await userEvent.click(screen.getByRole('checkbox', { name: 'Banana' }));
      expect(spy).toHaveBeenCalledWith(['Apple', 'Banana']);
    });

    it('triggers the onValues callback when values are removed', async () => {
      const spy = jest.fn();
      render(
        <Test
          options={options}
          onValues={spy}
          values={['Banana']}
        />,
      );
      await userEvent.click(screen.getByRole('checkbox', { name: 'Banana' }));
      expect(spy).toHaveBeenCalledWith([]);
    });

    it('updates when the value changes', async () => {
      render(<Test options={options} />);
      await userEvent.click(screen.getByRole('checkbox', { name: 'Banana' }));
      await userEvent.click(screen.getByRole('checkbox', { name: 'Apple' }));
      expect(screen.getAllByRole('checkbox', { checked: true })).toEqual([
        screen.getByRole('checkbox', { name: 'Apple' }),
        screen.getByRole('checkbox', { name: 'Banana' }),
      ]);
    });
  });

  describe('options as array of numbers', () => {
    const options = [0, 1, 2, 3];

    it('updates when the value changes', async () => {
      render(<Test options={options} />);
      await userEvent.click(screen.getByRole('checkbox', { name: '1' }));
      await userEvent.click(screen.getByRole('checkbox', { name: '2' }));
      expect(screen.getAllByRole('checkbox', { checked: true })).toEqual([
        screen.getByRole('checkbox', { name: '1' }),
        screen.getByRole('checkbox', { name: '2' }),
      ]);
    });
  });

  describe('options as array of objects', () => {
    describe('label', () => {
      const options = [
        { label: 'Apple' },
        { label: 'Banana' },
        { label: 'Orange' },
      ];

      it('updates when the value changes', async () => {
        render(<Test options={options} />);
        await userEvent.click(screen.getByRole('checkbox', { name: 'Banana' }));
        await userEvent.click(screen.getByRole('checkbox', { name: 'Orange' }));
        expect(screen.getAllByRole('checkbox', { checked: true })).toEqual([
          screen.getByRole('checkbox', { name: 'Banana' }),
          screen.getByRole('checkbox', { name: 'Orange' }),
        ]);
      });
    });

    describe('value', () => {
      it('is used as a options identity', async () => {
        const options = [
          { label: 'foo', value: 1 },
          { label: 'foo', value: 2 },
          { label: 'foo', value: 3 },
        ];
        render(
          <Test
            options={options}
            values={[2]}
          />,
        );
        expect(screen.getByRole('checkbox', { checked: true }).value).toEqual(
          '2',
        );
        const checkboxes = screen.getAllByRole('checkbox');
        await userEvent.click(checkboxes[0]);
        expect(screen.getAllByRole('checkbox', { checked: true })).toEqual([
          checkboxes[0],
          checkboxes[1],
        ]);
      });
    });

    describe('id', () => {
      it('is used as a options identity', async () => {
        const options = [
          { label: 'foo', id: 1 },
          { label: 'foo', id: 2 },
          { label: 'foo', id: 3 },
        ];
        render(
          <Test
            options={options}
            values={[2]}
          />,
        );
        expect(screen.getByRole('checkbox', { checked: true }).value).toEqual(
          '2',
        );
        const checkboxes = screen.getAllByRole('checkbox');
        await userEvent.click(checkboxes[0]);
        expect(screen.getAllByRole('checkbox', { checked: true })).toEqual([
          checkboxes[0],
          checkboxes[1],
        ]);
      });
    });

    describe('disabled', () => {
      it('sets the disabled attribute', () => {
        const options = [{ label: 'foo', disabled: true }];
        render(<Test options={options} />);
        expect(screen.getByRole('checkbox')).toBeDisabled();
      });

      it('will display a disabled checkboz as checked', () => {
        const options = [{ label: 'foo', disabled: true }];
        render(
          <Test
            options={options}
            values={['foo']}
          />,
        );
        expect(screen.getByRole('checkbox')).toBeChecked();
      });
    });

    describe('description', () => {
      it('sets the description', () => {
        const options = [{ label: 'foo', description: 'foo bar' }];
        render(<Test options={options} />);
        expect(screen.getByRole('checkbox')).toHaveAccessibleDescription(
          'foo bar',
        );
      });
    });

    describe('hint', () => {
      it('is used as a fallback for description', () => {
        const options = [{ label: 'foo', hint: 'foo bar' }];
        render(<Test options={options} />);
        expect(screen.getByRole('checkbox')).toHaveAccessibleDescription(
          'foo bar',
        );
      });

      it('is not used if description is present', () => {
        const options = [
          { label: 'foo', description: 'fizz buzz', hint: 'foo bar' },
        ];
        render(<Test options={options} />);
        expect(screen.getByRole('checkbox')).toHaveAccessibleDescription(
          'fizz buzz',
        );
      });
    });

    describe('html', () => {
      it('sets attributes on the checkbox', () => {
        const options = [
          { label: 'foo', html: { 'data-foo': 'bar', className: 'class' } },
        ];
        render(<Test options={options} />);
        expect(screen.getByRole('checkbox')).toHaveAttribute('data-foo', 'bar');
        expect(screen.getByRole('checkbox')).toHaveClass('class');
      });

      describe('ref', () => {
        it('will use a callback ref', () => {
          const ref = jest.fn();
          const options = [{ label: 'foo', html: { ref } }];
          render(<Test options={options} />);
          expect(ref).toHaveBeenCalledWith(expect.any(HTMLInputElement));
        });

        it('will use an object ref', () => {
          const ref = createRef();
          const options = [{ label: 'foo', html: { ref } }];
          render(<Test options={options} />);
          expect(ref.current).toBeInstanceOf(HTMLInputElement);
        });
      });
    });

    describe('group', () => {
      const options = [
        { label: 'Apple' },
        { label: 'Orange', group: 'Citrus' },
        { label: 'Lemon', group: 'Citrus' },
        { label: 'Raspberry', group: 'Berry' },
        { label: 'Strawberry', group: 'Berry' },
      ];

      it('renders grouped options', () => {
        const { container } = render(<Test options={options} />);
        expect(container).toMatchSnapshot();
      });

      it('selects an option', async () => {
        render(<Test options={options} />);
        const option = screen.getByRole('checkbox', { name: 'Citrus Lemon' });
        await userEvent.click(option);
        expect(screen.getByRole('checkbox', { checked: true })).toEqual(option);
      });
    });

    describe('other attributes', () => {
      it('does not render them', () => {
        const options = [{ label: 'foo', 'data-foo': 'bar' }];
        render(<Test options={options} />);
        expect(screen.getByRole('checkbox')).not.toHaveAttribute(
          'data-foo',
          'bar',
        );
      });
    });
  });

  describe('mapOption', () => {
    describe('mapOption returns an object', () => {
      const options = [
        { name: 'Apple' },
        { name: 'Banana' },
        { name: 'Orange' },
      ];

      it('maps options', async () => {
        render(
          <Test
            options={options}
            mapOption={({ name }) => ({ label: name })}
          />,
        );
        const option = screen.getByRole('checkbox', { name: 'Banana' });
        await userEvent.click(option);
        expect(screen.getByRole('checkbox', { checked: true })).toEqual(option);
      });

      it('accepts value as a primitive', () => {
        render(
          <Test
            options={options}
            values={['Banana']}
            mapOption={({ name }) => ({ label: name })}
          />,
        );
        expect(
          screen.getByRole('checkbox', { name: 'Banana', checked: true }),
        ).toBeInTheDocument();
      });
    });

    describe('mapOption returns a string', () => {
      const options = [
        { name: 'Apple' },
        { name: 'Banana' },
        { name: 'Orange' },
      ];

      it('maps options', async () => {
        render(
          <Test
            options={options}
            mapOption={({ name }) => name}
          />,
        );
        const option = screen.getByRole('checkbox', { name: 'Banana' });
        await userEvent.click(option);
        expect(screen.getByRole('checkbox', { checked: true })).toEqual(option);
      });

      it('accepts value as a primitive', () => {
        render(
          <Test
            options={options}
            values={['Banana']}
            mapOption={({ name }) => name}
          />,
        );
        expect(
          screen.getByRole('checkbox', { name: 'Banana', checked: true }),
        ).toBeInTheDocument();
      });
    });
  });
});

describe('name', () => {
  it('sets the name', () => {
    const options = ['Apple'];
    render(
      <Test
        options={options}
        name="foo"
      />,
    );
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toHaveProperty('name', 'foo');
  });
});

describe('onChange', () => {
  it('is called if a checkbox changes checked state', async () => {
    const options = ['Apple', 'Banana'];
    const onChange = jest.fn();
    render(
      <Test
        options={options}
        onChange={onChange}
        onValue={null}
      />,
    );
    const checkbox = screen.getByRole('checkbox', { name: 'Banana' });
    await userEvent.click(checkbox);
    expect(onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: checkbox,
      }),
    );
  });
});

describe('placeholderOption', () => {
  it('does not create an option', () => {
    const options = ['Apple', 'Banana'];
    render(
      <Test
        options={options}
        placeholderOption="Foo"
      />,
    );
    expect(
      screen.queryByRole('checkbox', { name: 'Foo' }),
    ).not.toBeInTheDocument();
  });
});

describe('classPrefix', () => {
  it('when null removes the class', () => {
    const options = ['Apple'];
    render(
      <Test
        options={options}
        classPrefix={null}
      />,
    );
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).not.toHaveClass();
  });

  it('sets the class', () => {
    const options = ['Apple'];
    render(
      <Test
        options={options}
        classPrefix="bar"
      />,
    );
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toHaveClass('bar__input');
  });
});

describe('renderWrapper', () => {
  it('customises the wrapper', () => {
    const spy = jest.fn(({ key, ...props }) => (
      <div
        key={key}
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={['Apple']}
        renderWrapper={spy}
        test="foo"
      />,
    );

    expect(screen.getByRole('checkbox').parentNode).toHaveAttribute(
      'data-foo',
      'bar',
    );
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderInput', () => {
  it('customises the input', () => {
    const spy = jest.fn((props) => (
      <input
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={['Apple']}
        renderInput={spy}
        test="foo"
      />,
    );

    expect(screen.getByRole('checkbox')).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderLabel', () => {
  it('customises the label', () => {
    const spy = jest.fn((props) => (
      <label
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={['Apple']}
        renderLabel={spy}
        test="foo"
      />,
    );

    expect(document.querySelector('label')).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderDescription', () => {
  it('customises the description', () => {
    const spy = jest.fn((props) => (
      <div
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={[{ label: 'Apple', description: 'fizz' }]}
        renderDescription={spy}
        test="foo"
      />,
    );

    const checkbox = screen.getByRole('checkbox');
    expect(
      document.getElementById(checkbox.getAttribute('aria-describedby')),
    ).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        option: expect.objectContaining({ label: 'Apple' }),
        group: undefined,
        checked: false,
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderGroup', () => {
  it('customises the group', () => {
    const spy = jest.fn(({ key, ...props }) => (
      <div
        key={key}
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={[{ label: 'Apple', group: 'fizz' }]}
        renderGroup={spy}
        test="foo"
      />,
    );

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox.parentNode.parentNode).toHaveAttribute('data-foo', 'bar');
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        group: expect.objectContaining({ label: 'fizz' }),
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});

describe('renderGroupLabel', () => {
  it('customises the group label', () => {
    const spy = jest.fn((props) => (
      <div
        data-foo="bar"
        {...props}
      />
    ));
    render(
      <Test
        options={[{ label: 'Apple', group: 'fizz' }]}
        renderGroupLabel={spy}
        test="foo"
      />,
    );

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox.parentNode.parentNode.firstElementChild).toHaveAttribute(
      'data-foo',
      'bar',
    );
    expect(spy).toHaveBeenCalledWith(
      expect.any(Object),
      {
        group: expect.objectContaining({ label: 'fizz' }),
      },
      expect.objectContaining({
        options: expect.any(Array),
        test: 'foo',
      }),
    );
  });
});
