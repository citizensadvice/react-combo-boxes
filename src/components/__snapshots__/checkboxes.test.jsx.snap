// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options as array of strings renders checkbox group 1`] = `
<div>
  <div
    class="react-combo-boxes-checkbox"
  >
    <input
      class="react-combo-boxes-checkbox__input"
      id="id_option_apple"
      name="name"
      type="checkbox"
      value="Apple"
    />
    <label
      class="react-combo-boxes-checkbox__label"
      for="id_option_apple"
    >
      Apple
    </label>
  </div>
  <div
    class="react-combo-boxes-checkbox"
  >
    <input
      class="react-combo-boxes-checkbox__input"
      id="id_option_banana"
      name="name"
      type="checkbox"
      value="Banana"
    />
    <label
      class="react-combo-boxes-checkbox__label"
      for="id_option_banana"
    >
      Banana
    </label>
  </div>
  <div
    class="react-combo-boxes-checkbox"
  >
    <input
      class="react-combo-boxes-checkbox__input"
      id="id_option_orange"
      name="name"
      type="checkbox"
      value="Orange"
    />
    <label
      class="react-combo-boxes-checkbox__label"
      for="id_option_orange"
    >
      Orange
    </label>
  </div>
</div>
`;

exports[`options options as array of objects group renders grouped options 1`] = `
<div>
  <div
    class="react-combo-boxes-checkbox"
  >
    <input
      class="react-combo-boxes-checkbox__input"
      id="id_option_apple"
      name="name"
      type="checkbox"
      value="Apple"
    />
    <label
      class="react-combo-boxes-checkbox__label"
      for="id_option_apple"
    >
      Apple
    </label>
  </div>
  <div
    class="react-combo-boxes-checkbox-group"
  >
    <div
      class="react-combo-boxes-checkbox-group__label"
    >
      Citrus
    </div>
    <div
      class="react-combo-boxes-checkbox"
    >
      <input
        class="react-combo-boxes-checkbox__input"
        id="id_option_orange"
        name="name"
        type="checkbox"
        value="Orange"
      />
      <label
        class="react-combo-boxes-checkbox__label"
        for="id_option_orange"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Citrus 
        </span>
        Orange
      </label>
    </div>
    <div
      class="react-combo-boxes-checkbox"
    >
      <input
        class="react-combo-boxes-checkbox__input"
        id="id_option_lemon"
        name="name"
        type="checkbox"
        value="Lemon"
      />
      <label
        class="react-combo-boxes-checkbox__label"
        for="id_option_lemon"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Citrus 
        </span>
        Lemon
      </label>
    </div>
  </div>
  <div
    class="react-combo-boxes-checkbox-group"
  >
    <div
      class="react-combo-boxes-checkbox-group__label"
    >
      Berry
    </div>
    <div
      class="react-combo-boxes-checkbox"
    >
      <input
        class="react-combo-boxes-checkbox__input"
        id="id_option_raspberry"
        name="name"
        type="checkbox"
        value="Raspberry"
      />
      <label
        class="react-combo-boxes-checkbox__label"
        for="id_option_raspberry"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Berry 
        </span>
        Raspberry
      </label>
    </div>
    <div
      class="react-combo-boxes-checkbox"
    >
      <input
        class="react-combo-boxes-checkbox__input"
        id="id_option_strawberry"
        name="name"
        type="checkbox"
        value="Strawberry"
      />
      <label
        class="react-combo-boxes-checkbox__label"
        for="id_option_strawberry"
      >
        <span
          class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
        >
          Berry 
        </span>
        Strawberry
      </label>
    </div>
  </div>
</div>
`;
